java-checks/src/main/java/org/sonar/java/checks/ConstantsShouldBeStaticFinalCheck.java

```diff
@@ -86,32 +85,36 @@ public void leaveNode(Tree tree) {
 
   private static boolean hasConstantInitializer(VariableTree variableTree) {
     ExpressionTree init = variableTree.initializer();
-    if (init != null) {
-      if (ExpressionUtils.skipParentheses(init).is(Tree.Kind.METHOD_REFERENCE)) {
-        MethodReferenceTree methodRef = (MethodReferenceTree) ExpressionUtils.skipParentheses(init);
-        if (isInstanceIdentifier(methodRef.expression())) {
-          return false;
-        }
-      }
-      if (init.is(Tree.Kind.NEW_ARRAY)) {
+    if (init == null) {
+      return false;
+    }
+
+    var deparenthesized = ExpressionUtils.skipParentheses(init);
+
+    if (deparenthesized.is(Tree.Kind.METHOD_REFERENCE)) {
+      MethodReferenceTree methodRef = (MethodReferenceTree) deparenthesized;
+      if (isInstanceIdentifier(methodRef.expression())) {
         return false;
       }
-      return !containsChildMatchingPredicate((JavaTree) init,
-        (tree -> isIgnoredKind(tree) || isThisOrSuper(tree)));
     }
-    return false;
+    return !containsChildMatchingPredicate((JavaTree) deparenthesized,
+      (ConstantsShouldBeStaticFinalCheck::isNonStaticOrFinal));
   }
 
-  private static boolean isIgnoredKind(Tree tree) {
-    return tree.is(Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
-  }
-
-  private static boolean isThisOrSuper(Tree tree) {
-    if (tree.is(Tree.Kind.IDENTIFIER)) {
-      String name = ((IdentifierTree) tree).name();
-      return "super".equals(name) || "this".equals(name);
-    }
-    return false;
+  private static boolean isNonStaticOrFinal(Tree tree) {
+    return switch (tree.kind()) {
+      case METHOD_INVOCATION, NEW_CLASS, NEW_ARRAY -> true;
```

Comment:
I just realized that there is one operation that returns a non-constant result even if both source operands are static final: array access. We can't make any assumptions over the contents even for static final arrays, so we need to add it to the non-static-or-final-case:

```suggestion
      case METHOD_INVOCATION, NEW_CLASS, NEW_ARRAY, ARRAY_ACCESS_EXPRESSION -> true;
```